kimchi/src/circuits/constraints.rs

```diff
@@ -136,129 +135,204 @@ pub struct ConstraintSystem<F: FftField> {
+                            col: wire.col,
+                            row: wire.row,
+                        },
+                    ));
+                }
+            }
+
+            // for public gates, only the left wire is toggled
+            if row < self.public && gate.coeffs[0] != F::one() {
+                return Err(GateError::IncorrectPublic(row));
+            }
+
+            // check the gate's satisfiability
+            gate.verify(row, &witness, self, public)
+                .map_err(|err| GateError::Custom { row, err })?;
         }
-        shift
+
+        // all good!
+        Ok(())
     }
 
-    /// Returns the field element that represents a position
-    fn cell_to_field(&self, &Wire { row, col }: &Wire) -> F {
-        self.map[col][row]
+    /// evaluate witness polynomials over domains
+    pub fn evaluate(&self, w: &[DP<F>; COLUMNS], z: &DP<F>) -> WitnessOverDomains<F> {
+        // compute shifted witness polynomials
+        let w8: [E<F, D<F>>; COLUMNS] =
+            array_init(|i| w[i].evaluate_over_domain_by_ref(self.domain.d8));
+        let z8 = z.evaluate_over_domain_by_ref(self.domain.d8);
+
+        let w4: [E<F, D<F>>; COLUMNS] = array_init(|i| {
+            E::<F, D<F>>::from_vec_and_domain(
+                (0..self.domain.d4.size)
+                    .map(|j| w8[i].evals[2 * j as usize])
+                    .collect(),
+                self.domain.d4,
+            )
+        });
+        let z4 = DP::<F>::zero().evaluate_over_domain_by_ref(D::<F>::new(1).unwrap());
+
+        WitnessOverDomains {
+            d4: WitnessShifts {
+                next: WitnessEvals {
+                    w: array_init(|i| w4[i].shift(4)),
+                    // TODO(mimoo): change z to an Option? Or maybe not, we might actually need this dummy evaluation in the aggregated evaluation proof
+                    z: z4.clone(), // dummy evaluation
+                },
+                this: WitnessEvals {
+                    w: w4,
+                    z: z4, // dummy evaluation
+                },
+            },
+            d8: WitnessShifts {
+                next: WitnessEvals {
+                    w: array_init(|i| w8[i].shift(8)),
+                    z: z8.shift(8),
+                },
+                this: WitnessEvals { w: w8, z: z8 },
+            },
+        }
     }
 }
 
-/// Represents an error found when verifying a witness with a gate
-#[derive(Debug)]
-pub enum GateError {
-    /// Some connected wires have different values
-    DisconnectedWires(Wire, Wire),
-    /// A public gate was incorrectly connected
-    IncorrectPublic(usize),
-    /// A specific gate did not verify correctly
-    Custom { row: usize, err: String },
-}
+impl<F: FftField + SquareRootField> Builder<F> {
+    /// Set up the number of public inputs.
+    /// If not invoked, it equals `0` by default.
+    pub fn public(mut self, public: usize) -> Self {
+        self.public = public;
+        self
+    }
```

Comment:
unrelated to the PR, but a thought: it'd be interesting if we had public input gates (or even public input and public output gates) so that we wouldn't have to pass that argument. The gates wouldn't really exist (they would still use the double generic gate underneath) but could allow nicer abstractions

Reply:
oh that is very cute, definitely a good idea

Reply:
I'm not sure that adding this at the `GateType` level is a good idea; then, we have to remember to handle this `PublicInput` kind everywhere, including where it appears part-way through the circuit. It might still make sense to add this to snarky / Izzy's circuit builder API, but I'm not sure whether it actually makes things cleaner than e.g. snarky's input `Typ.t` interface.